### PR TITLE
refactor: convert Jenkinsfile to declarative pipeline syntax

### PR DESCRIPTION
**Pull request by jdrueckert**

### CODE REVIEW

**Comment by jdrueckert** [comment]
[Fix Checkstyle Findings](https://trello.com/c/SvpELrak/121-fix-checkstyle-findings)

File: Jenkinsfile
Comment (keturn) [inline Jenkinsfile]: thought: this stuff looks like it's all from the code that was in the setup step, with some small edits to give conditions variable names. 👍

File: Jenkinsfile
Comment (keturn) [inline Jenkinsfile:73]: suggestion: this does look equivalent to the old version, but ideally we'd publish that htmlDependencyReport even if the jar task fails

File: Jenkinsfile
Comment (keturn) [inline Jenkinsfile:101]: Thinking about whether we want `allowEmptyResults: true` here. I guess some modules don't define tests at all; what happens when we run `unitTest` there? If they don't produce any test-results, then we'd need to allow empty results.

But if they do produce test-results (that just say "0 tests discovered") then we don't have any reason to allow empty results.

File: Jenkinsfile
Comment (keturn) [inline Jenkinsfile]: rename `specialBranch` to `publishedBranch`, maybe?

File: Jenkinsfile
Comment (keturn) [inline Jenkinsfile]: 👍 looks like this is what we learned from the other Jenkinsfile. `'` style for things that include credentials, and double-backslash to spread shell command arguments out over multiple lines.

File: Jenkinsfile
Comment (keturn) [inline Jenkinsfile]: another case where we'll want the `recordIssues` moved to a `post` step so that we record what we have even if a check fails.

File: Jenkinsfile
Comment (keturn) [inline Jenkinsfile:155]: can this `fileExists` thing go inside a `when` expression instead of `script`?

File: Jenkinsfile
Comment (keturn) [inline Jenkinsfile]: 🚦 This is something we'll want to try to test.

I don't _expect_ an empty `failure {}` block here to alter the outcome one way or the other, but I am not sure.

**Review by keturn — COMMENTED** [review]
Looks reasonable. Ideally we'll figure out how to run it before merging.

There are some things about how we handle failure cases that we'll want to change, but I'd be okay with leaving those changes for a separate PR if you want to keep this one strictly about the scripted/declarative syntax.

**Comment by Cervator** [comment]
There are more related changes over in https://github.com/Nanoware/ModuleJteConfig/blob/develop/Jenkinsfile right now, but also an apparent dependency on upgrading Jenkins and some plugins before Declarative will play along nicely with JTE.

**Comment by keturn** [comment]
Specifically, it's the "Pipeline: Model API" plugin that needs to be upgraded, as supporting declarative pipeline with Jenkins Templating Engine is in the next major revision. And for whatever reason, it's not something we seem to be able to upgrade standalone. So I concur, it looks like we need to upgrade the whole Jenkins controller.

**Comment by keturn** [comment]
but I also am quite disappointed with the error reporting and troubleshooting for this error, which led me to think: Do we need JTE for this?

I don't think we're making use of its templating abilities here. If I understand correctly, we just want to be able to define the pipeline in a way that a module committer cannot override by adding their own Jenkinsfile.

I found a couple possibilities but none quite fit:
* Cloudbees lets us define the job with a "Custom script," but that script is stored with the Jenkins configuration, not as a version-controlled file.
* The [Multibranch Defaults](https://github.com/jenkinsci/pipeline-multibranch-defaults-plugin) plugin does something very similar without a Cloudbees license.
* [Config-Driven Pipeline](https://plugins.jenkins.io/config-driven-pipeline/) is better in that it lets us version-control the Jenkinsfile, but it relies on a config file in the project repository. It would probably fit our use case if we implemented https://github.com/jenkinsci/config-driven-pipeline-plugin/issues/25

and as difficult as it was for me to see how it was the use of JTE that made for our failure case here, I don't really have any reason to believe that a similar failure in any of those other plugins would be any less obtuse. 😐

File: Jenkinsfile
Comment (keturn) [reply Jenkinsfile]: removed this, as Jenkins was complaining that a empty block was a syntax error

**Comment by keturn** [comment]
we upgraded Jenkins and the builds are closer to working now, or at least giving more relevant error messages.

File: Jenkinsfile
Comment (jdrueckert) [reply Jenkinsfile:101]: I _think_ they don't produce any test-results if there are no tests :thinking: not entirely sure though

File: Jenkinsfile
Comment (jdrueckert) [reply Jenkinsfile]: `published` sounds like it already was published, but we're only about to publish it respectively trying to decide whether or not to publish it... :thinking: 
maybe `publishBranch` or as it's a boolean and not a branch name, maybe even only `shouldPublish`?

File: Jenkinsfile
Comment (jdrueckert) [reply Jenkinsfile:155]: afaik it cannot, as we need to execute something before we can check and `when` IIRC is only executed at the very start of / before a stage

File: Jenkinsfile
Comment (Cervator) [inline Jenkinsfile]: Minor typo :-) Enough to make build go boom.
```suggestion
String realProjectName = jobNameParts.length < 2 ? env.JOB_NAME : jobNameParts[jobNameParts.length - 2]
```

File: Jenkinsfile
Comment (Cervator) [inline Jenkinsfile:8]: If the X here is meant to tie it to https://jenkins.terasology.io/teraorg/job/Nanoware/job/TerasologyModules/job/X/ maybe include a link to it? This could get confusing over time if things change and only part of the picture is readily visible :-)

File: Jenkinsfile
Comment (Cervator) [reply Jenkinsfile]: +1 for `shouldPublish` :-)

Or maybe `publishableBranch`

**Review by Cervator — CHANGES_REQUESTED** [review]
Tiny typo fix might be enough! Looks nice otherwise :-)

**Review by Cervator — APPROVED** [review]

File: Jenkinsfile
Comment (jdrueckert) [reply Jenkinsfile:8]: like so?

**Review by keturn — APPROVED** [review]
Goal of "convert to declarative" achieved, with some good improvements for clarity along the way. Result seems to be working now that we have learned how to use the Nanoware jobs to confirm.

Ship it!

**Pull request merged**
